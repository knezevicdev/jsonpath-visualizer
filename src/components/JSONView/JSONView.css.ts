import styled from 'styled-components';

export const Wrapper = styled.div`
  width: 100%;
  height: 90vh;
  background-color: #272822;
  overflow: auto;
`;
