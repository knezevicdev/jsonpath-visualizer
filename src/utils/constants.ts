export const ARRAY_GROUP_LIMIT = 100;
