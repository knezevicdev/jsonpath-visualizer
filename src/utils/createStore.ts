export type TJson = null | Record<string, unknown> | any[];

export const createStore = () => {
  return {
    json: null as TJson,
    jsonPath: '' as string,
  };
};

export type TStore = ReturnType<typeof createStore>;
