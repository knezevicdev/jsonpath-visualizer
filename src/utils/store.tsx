import React, { ComponentType, ReactPropTypes } from 'react';
import { createStore, TStore } from './createStore';
import { useLocalStore } from 'mobx-react';

const storeContext = React.createContext<TStore | null>(null);

export const StoreProvider: React.FC = ({ children }) => {
  const store = useLocalStore(createStore);
  return <storeContext.Provider value={store}>{children}</storeContext.Provider>;
};

export const withStore = (WrappedComponent: ComponentType): React.FC => {
  const StoreHOC: React.FC = (props) => {
    return (
      <StoreProvider>
        <WrappedComponent {...props} />
      </StoreProvider>
    );
  };

  return StoreHOC;
};

export const useStore = () => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
