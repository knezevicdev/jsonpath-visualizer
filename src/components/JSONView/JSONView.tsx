import React from 'react';
import { useStore } from 'utils/store';
import { observer } from 'mobx-react';

const JSONView: React.FC = () => {
  const store = useStore();
  return (
    <div>
      <pre>{JSON.stringify(store.json, null, 2)}</pre>
    </div>
  );
};

export default observer(JSONView);
