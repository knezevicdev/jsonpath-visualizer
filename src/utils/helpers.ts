import { transform as _transform, isObject as _isObject, mapKeys as _mapKeys, assign as _assign } from 'lodash';
import { JSONPath } from 'jsonpath-plus';

import { TJson } from './store';

// eslint-disable-next-line
export const toType = (obj: any): string => {
  let type = getType(obj);
  if (type === 'number') {
    if (isNaN(obj)) {
      type = 'nan';
    } else if ((obj | 0) != obj) {
      //bitwise OR produces integers
      type = 'float';
    } else {
      type = 'integer';
    }
  }
  return type;
};

//source: http://stackoverflow.com/questions/7390426/better-way-to-get-type-of-a-javascript-variable/7390612#7390612
const getType = (obj: any): string => {
  return {}.toString
    .call(obj)
    .match(/\s([a-zA-Z]+)/)[1]
    .toLowerCase();
};

// eslint-disable-next-line
export const objectPaths = (obj: any): Record<string, unknown> => {
  return _transform(
    obj,
    (result: Record<string, unknown>, value: unknown, key: string) => {
      if (_isObject(value)) {
        const flatMap = _mapKeys(objectPaths(value), function (_mvalue, mkey) {
          return `${key}.${mkey}`;
        });

        _assign(result, flatMap);
      } else {
        result[String(key)] = false;
      }

      return result;
    },
    {},
  );
};

export const matchPaths = (json: TJson, path: string): string[] => {
  const paths = JSONPath({
    path,
    json,
    resultType: 'pointer',
  }).map((pointer: string) => pointer.replace(/\//g, '.').substring(1));

  if (paths.includes('')) {
    return [''];
  }

  return paths.reduce((matchedPaths: string[], path: string) => {
    for (const matchedPath in matchedPaths) {
      if (path.includes(matchedPath)) {
        return matchedPaths;
      }
    }

    return [...matchedPaths, path];
  }, []);
};
