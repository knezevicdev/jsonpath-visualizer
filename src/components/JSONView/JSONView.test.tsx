import React from 'react';
import { shallow } from 'enzyme';

import JSONView from './JSONView';
import { StoreProvider } from 'utils/store';

describe('JSONView', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(
      <StoreProvider>
        <JSONView />
      </StoreProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
