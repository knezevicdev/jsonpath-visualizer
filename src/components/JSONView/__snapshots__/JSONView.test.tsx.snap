// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONView renders snapshots 1`] = `
<ContextProvider
  value={
    Object {
      "json": null,
      "jsonPath": "",
    }
  }
>
  <JSONView />
</ContextProvider>
`;
